source/code/omi_preexec/mycnf-scanner.cpp

```diff
@@ -0,0 +1,251 @@
+
+#include "mycnf-scanner.h"
+
+namespace MySQL
+{
+    const std::wstring cIncludeDirKeyword = L"!includedir";
+    const std::wstring cIncludeKeyword = L"!include";
+
+    /**
+     Given a line from My SQL configuration file containing an include* directive,
+     this function extracts the path used in the directive.
+     
+     MySQL extracts the path with the following algorithm:
+     Path begins at the first non-space character after the keyword. It continues till 
+     the end of the line, while ignoring the trailing whitespacelast non-space character.
+     (Reference: get_argument function in my_default.cc in mysql-server source)
+
+         !include(dir)?[\s]+path[\s]*
+
+     \param[in] file  containing the directive(used only for reporting errors).
+     \param[in] line  from the config file that has already been trimmed.
+     \param[in] keywordlen length of the directive.
+     \returns   SCXFilePath representing the path
+     \throws     MySQL_ConfigScannerException  if the line contains does not follow the syntax
+     */
+    static inline SCXCoreLib::SCXFilePath GetPathFromDirective(const SCXCoreLib::SCXFilePath& file, const std::wstring& line, size_t keywordLen)
+    {
+        if (line.size() > keywordLen)
+        {
+            std::wstring path = SCXCoreLib::StrTrim(line.substr(keywordLen, line.size()- keywordLen));
+            if ( path.size() > 0 )
+            {
+                return SCXCoreLib::SCXFilePath(path);
+            }
+        }
+
+        std::wstringstream ss;
+        ss << L"invalid directive in file, " << file.Get() << L" : " << line;
+        throw MySQL_ConfigScannerException(ss.str(), SCXSRCLOCATION);
+    }
+
+    /**
+       Scan a MySQL configuration file looking for 'port' and 'bind-address'
+       options, and add an entry to the given auth file for each 'port'/
+       'bind-address' pair. For '!include' directive, calls itself. 
+       and '!includedir' directive, LoadCnfDirectory is called.
+
+       \param[in]  cnfFile                       path of the MySQL configuration file
+       \param[in]  auth                          authentication object that has been 'Load'ed.
+       \throws     SCXFilePathNotFoundException  if cnfFile (or any of the paths pointed by "!include" directives) is not found
+       \throws     MySQL_ConfigScannerException  if the include directives form a cycle
+    */
+
+    void MySQL_ConfigurationScanner::LoadCnfFile(
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" loading: ", file.Get()));
+        std::pair<std::set< std::wstring >::iterator, bool> ins = m_visited.insert(file.Get());
+        if (!ins.second)
+        {
+            std::wstringstream ss;
+            ss << file.Get() << L"is used in !include more than once.";
+            throw MySQL_ConfigScannerException(ss.str(), SCXSRCLOCATION);
```

Comment:
Using generic exceptions like this is a double edged sword:

Pro: It's easy, one exception serves all,
Con: Impossible to easily catch any specific exception for special handling

Just FYI. I presume you don't need to catch specific exceptions, so you probably don't care.